lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+using Newtonsoft.Json;
+
+namespace Org.Apache.REEF.Common.Telemetry
+{
+    /// <summary>
+    /// This class maintains a collection of the data for all the metrics for metrics service. 
+    /// When new metric data is received, the data in the collection will be updated.
+    /// After the data is processed, the changes since last process will be reset.
+    /// </summary>
+    public sealed class MetricsData : IMetrics
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
+
+        JsonSerializerSettings settings = new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All
+        };
+
+        /// <summary>
+        /// Registration of metrics
+        /// </summary>
+        private ConcurrentDictionary<string, MetricTracker> _metricsMap = new ConcurrentDictionary<string, MetricTracker>();
+
+        [Inject]
+        internal MetricsData()
+        {
+        }
+
+        /// <summary>
+        /// Deserialization.
+        /// </summary>
+        /// <param name="serializedMetricsString">string to Deserialize.</param>
+        [JsonConstructor]
+        internal MetricsData(string serializedMetricsString)
+        {
+            var metrics = JsonConvert.DeserializeObject<IList<MetricTracker>>(serializedMetricsString, settings);
+
+            foreach (var m in metrics)
+            {
+                _metricsMap.TryAdd(m.GetMetric().Name, m);
+            }
+        }
+
+        public void RegisterMetric(IMetric metric)
+        {
+            if (!_metricsMap.TryAdd(metric.Name, new MetricTracker(metric)))
+            {
+                throw new ArgumentException("The metric [{0}] already exists.", metric.Name);
+            }
+        }
+
+        public bool TryGetMetric(string name, out IMetric me)
+        {
+            if (!_metricsMap.TryGetValue(name, out MetricTracker tracker))
+            {
+                me = null;
+                return false;
+            }
+            me = tracker.GetMetric();
+            return true;
+        }
+
+        public IEnumerable<MetricTracker> GetMetricTrackers()
+        {
+            return _metricsMap.Values;
+        }
+
+        /// <summary>
+        /// Flushes changes since last sink for each metric. 
+        /// Called when Driver is sinking metrics.
+        /// </summary>
+        /// <returns>Key value pairs of metric name and record that was flushed.</returns>
+        public IEnumerable<KeyValuePair<string, MetricTracker.MetricRecord>> FlushMetricRecords()
+        {
+            // for each metric, flush the records and create key value pairs
+            return _metricsMap.SelectMany(kv => kv.Value.FlushChangesSinceLastSink().Select(r => new KeyValuePair<string, MetricTracker.MetricRecord>(kv.Key, r)));
```

Comment:
here and everywhere: the line is too long. we try to keep your lines no longer than 120 characters; in most cases, it's below 100.